target() cannot be called outside a hypothesis test

I have a test `assert_allclose_phase` which checks whether two arrays are almost equal modulo one. When running under `hypothesis`, this seems like a sensible place to put `target`, albeit with a(n optional?) `name` argument in case I want to call it twice in the same test. Unfortunately if I do so then I can't use the same function in non-`hypothesis` tests:

```
E           hypothesis.errors.InvalidArgument: Calling target() outside of a test is invalid.
```

My current workaround is to call `target` only if `name` is provided, but this clutters simple tests that wouldn't need a `name`.

--- a/hypothesis-python/RELEASE.rst
+++ b/hypothesis-python/RELEASE.rst
@@ -1,0 +1,9 @@
+RELEASE_TYPE: minor
+
+This release adds :func:`hypothesis.currently_in_test_context`, which can be used
+to check whether the calling code is currently running inside an
+:func:`@given <hypothesis.given>` or :doc:`stateful <stateful>` test.
+
+This is most useful for third-party integrations and assertion helpers which may
+wish to use :func:`~hypothesis.assume` or :func:`~hypothesis.target`, without also
+requiring that the helper only be used from property-based tests (:issue:`2581`).

--- a/hypothesis-python/docs/strategies.rst
+++ b/hypothesis-python/docs/strategies.rst
@@ -101,7 +101,12 @@ features (if you need them) and better long-term guarantees about maintenance.
 We particularly encourage pull requests for new composable primitives that
 make implementing other strategies easier, or for widely used types in the
 standard library. Strategies for other things are also welcome; anything with
-external dependencies just goes in hypothesis.extra.
+external dependencies just goes in ``hypothesis.extra``.
+
+Tools such as assertion helpers may also need to check whether the current
+test is using Hypothesis:
+
+.. autofunction:: hypothesis.currently_in_test_context
 
 
 .. _entry-points:

--- a/hypothesis-python/src/hypothesis/__init__.py
+++ b/hypothesis-python/src/hypothesis/__init__.py
@@ -22,7 +22,14 @@ failing examples it finds.
 
 import hypothesis._error_if_old  # noqa  # imported for side-effect of nice error
 from hypothesis._settings import HealthCheck, Phase, Verbosity, settings
-from hypothesis.control import assume, event, note, reject, target
+from hypothesis.control import (
+    assume,
+    currently_in_test_context,
+    event,
+    note,
+    reject,
+    target,
+)
 from hypothesis.core import example, find, given, reproduce_failure, seed
 from hypothesis.entry_points import run
 from hypothesis.internal.entropy import register_random
@@ -30,21 +37,22 @@ from hypothesis.utils.conventions import infer
 from hypothesis.version import __version__, __version_info__
 
 __all__ = [
-    "settings",
-    "Verbosity",
     "HealthCheck",
     "Phase",
+    "Verbosity",
     "assume",
-    "reject",
-    "seed",
-    "given",
-    "reproduce_failure",
-    "find",
-    "example",
-    "note",
+    "currently_in_test_context",
     "event",
+    "example",
+    "find",
+    "given",
     "infer",
+    "note",
     "register_random",
+    "reject",
+    "reproduce_failure",
+    "seed",
+    "settings",
     "target",
     "__version__",
     "__version_info__",

--- a/hypothesis-python/src/hypothesis/control.py
+++ b/hypothesis-python/src/hypothesis/control.py
@@ -45,6 +45,18 @@ def assume(condition: object) -> bool:
 _current_build_context = DynamicVariable(None)
 
 
+def currently_in_test_context() -> bool:
+    """Return ``True`` if the calling code is currently running inside an
+    :func:`@given <hypothesis.given>` or :doc:`stateful <stateful>` test,
+    ``False`` otherwise.
+
+    This is useful for third-party integrations and assertion helpers which
+    may be called from traditional or property-based tests, but can only use
+    :func:`~hypothesis.assume` or :func:`~hypothesis.target` in the latter case.
+    """
+    return _current_build_context.value is not None
+
+
 def current_build_context():
     context = _current_build_context.value
     if context is None:

--- a/hypothesis-python/tests/cover/test_control.py
+++ b/hypothesis-python/tests/cover/test_control.py
@@ -21,11 +21,13 @@ from hypothesis.control import (
     _current_build_context,
     cleanup,
     current_build_context,
+    currently_in_test_context,
     event,
     note,
 )
 from hypothesis.errors import CleanupFailed, InvalidArgument
 from hypothesis.internal.conjecture.data import ConjectureData as TD
+from hypothesis.stateful import RuleBasedStateMachine, rule
 from hypothesis.strategies import integers
 from tests.common.utils import capture_out
 
@@ -156,3 +158,21 @@ def test_prints_all_notes_in_verbose_mode():
     v = out.getvalue()
     for x in sorted(messages):
         assert x in v
+
+
+def test_not_currently_in_hypothesis():
+    assert currently_in_test_context() is False
+
+
+@given(integers())
+def test_currently_in_hypothesis(_):
+    assert currently_in_test_context() is True
+
+
+class ContextMachine(RuleBasedStateMachine):
+    @rule()
+    def step(self):
+        assert currently_in_test_context() is True
+
+
+test_currently_in_stateful_test = ContextMachine.TestCase

